### PR TITLE
Fix checkboxes in Firefox and IE/Edge

### PR DESCRIPTION
**Pull request by patrickkettner**
firefox doesn't allow for the input elements style to be completely overwritten
like it has been in chrome. So instead I am hiding the checkbox via opacity and
loading the checkbox as the background of the label. This has the nice side
effect of fixing an issue with Edge and svg backgrounds on input elements.
Finally, firefox is choking on `#` in svg data uris. escaping the url fixes
that.

closes tastejs/todomvc#1408

[Review on Reviewable](https://reviewable.io/reviews/tastejs/todomvc-app-css/8)

### CODE REVIEW

File: index.css [hunk 1]
```diff
@@ -117,28 +117,32 @@ input[type="checkbox"] {
...
-	display: none;
+.toggle-all {
+	text-align: center;
+	border: none; /* Mobile Safari */
+	opacity: 0;
+	position: absolute;
 }
 
-.toggle-all {
+.toggle-all + label{
```
Comment (sindresorhus) [inline index.css]:
```
.toggle-all + label {
```

**Comment by sindresorhus** [comment]
> Finally, firefox is choking on # in svg data uris. escaping the url fixes
that.

I'd prefer we not do this. It makes the SVG totally unreadable. Is there a Firefox bug ticket for this?

**Comment by sindresorhus** [comment]
Nice work @patrickkettner! :)

**Comment by samccone** [comment]
Also will close https://github.com/tastejs/todomvc/issues/1259

thanks a million @patrickkettner :+1: :+1: :+1: :+1: :+1:  :rocket:

**Comment by patrickkettner** [comment]
@sindresorhus just escaping the `#` works too. I reduced it to just that bit. Not 100% intuitive (should inline comments be added?). Couldn't find a relevant FF bug, but I planned on opening one this afternoon.

**Comment by passy** [comment]
@patrickkettner This looks great! But it seems that the `toggle-all` checkbox is no longer clickable after the change.

**Comment by patrickkettner** [comment]
@passy hmmm... it is for me. What browser are and framework are you testing it in?

**Comment by sindresorhus** [comment]
ping @passy

**Comment by samccone** [comment]
pong @passy :wink:

**Comment by passy** [comment]
@patrickkettner 

Browsers tested:

- Chrome 44.0.2403.130 on Linux
- Firefox 41.0a2 (2015-08-03) on Linux
- Chrome 45.0.2454.62 on Android
- Safari 8.0.8 on OS X

**Comment by patrickkettner** [comment]
Pardon?

**Comment by passy** [comment]
Sorry, these are all the browsers I've tried and managed to reproduce it in.

**Comment by patrickkettner** [comment]
no worries, I actually tried them on all of browserstacks desktop browsers and didn't spot any issues. Haven't tried mobile, however

**Comment by arthurvr** [comment]
I am able to repro the same issues as @passy on Chrome 46 on OS X.

**Comment by patrickkettner** [comment]
thanks, @arthurvr. ill check it out shortly

**Comment by patrickkettner** [comment]
which projects are you testing this against?

On Sun, Aug 23, 2015 at 2:22 PM, Arthur Verschaeve <notifications@github.com
> wrote:

> I am able to repro the same issues as @passy <https://github.com/passy>
> on Chrome 46 on OS X.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tastejs/todomvc-app-css/pull/8#issuecomment-133892461>
> .
>



-- 
patrick

**Comment by sindresorhus** [comment]
ping @arthurvr @passy

**Comment by passy** [comment]
@patrickkettner I think I used `typescript-backbone` because it uses the new CSS structure.

**Comment by patrickkettner** [comment]
So the reason it is broken on that particular one is because the label/input connection is broken. Its using a `for` attribute that is not associated with any `id` or `name`.  I would view this as a bug in the html, but I can make adjustments on the css end if y'all think this is the right thing to do

**Comment by patrickkettner** [comment]
@passy @sindresorhus So the reason it is broken on that particular one is because the label/input connection is broken. Its using a for attribute that is not associated with any id or name. I would view this as a bug in the html, but I can make adjustments on the css end if y'all think this is the right thing to do

**Comment by passy** [comment]
I'm pretty sure I've tried three in total. If an HTML change is required I
think that's alright. Which app does it work with unmodified so I can try
it as it is?

**Comment by samccone** [comment]
>  it is broken on that particular one is because the label/input connection is broken

^^ makes sense, imho we can just manually go in and introduce a line or two of CSS to resolve this per framework

**Comment by patrickkettner** [comment]
@passy I am not sure where I was testing it before, but if you add `id=toggle-all` or `name=toggle-all` to the markup on the typescript-backbone example you will see it working

@samccone just to clarify, this is something that should be fixed in HTML, rather than CSS (in my opinion, that is)

**Comment by samccone** [comment]
Unfortunately some of the example frameworks have limitations around the DOM structure, but we will just need to enumerate all of the broken ones and list out how we will fix them.

**Comment by patrickkettner** [comment]
there are frameworks that don't allow for an id?

**Comment by patrickkettner** [comment]
Ping @samccone @passy

**Comment by sindresorhus** [comment]
@patrickkettner Yes, let's try to get this merged, finally.

> just escaping the # works too. I reduced it to just that bit. Not 100% intuitive (should inline comments be added?). Couldn't find a relevant FF bug, but I planned on opening one this afternoon.

There needs to be a link to an existing or new FF bug ticket as an inline comment.

---

I tried this change with the Backbone app and compared with existing CSS and noticed a few incorrect changes:

- The toggle all button moved. Make sure it's in the exact same spot.
- Long lines are no longer wrapped.  screen shot 2017-01-24 at 11 59 01

---

Someone needs to test this in IE and latest Edge and confirm it's working fine there. // @jdalton

---

As for the broken toggle-all connection, we can just make this a major version in this package, and ensure we also update https://github.com/tastejs/todomvc-app-template/blob/master/index.html (Submitting a PR there with a fix would help move this along)

**Comment by patrickkettner** [comment]
@sindresorhus thank you for taking the time to look at this

so, ironically, reducing the escaped value for Firefox broken it for Edge and IE. They _require_ the full value to be escaped. There is an [open bug](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7157459/) to fix this, which I have marked as highly important internally due to todo's reliance (amongst others) on it. As a result I reverted the reduced escaping back to the full escape.
I also created a firefox bug for the hash issue, which is also commented on in the css.

The long lines were a type on my part ("pre" !== "pre-wrap"), which was later changed in master by using `word-break: break-all`. It is now also fixed

I fixed the placement of the toggle all button by tweaking the absolute positioning.

As for the toggle-all not being clickable, I am not able to reproduce it. If it _is_ reproducible anywhere, it would be due to that demo having broken html (which I would be happy to PR a fix for, if anyone finds it).

Let me know if there is anything else I can do to get this merged!

**Comment by patrickkettner** [comment]
ping @samccone @sindresorhus

**Comment by sindresorhus** [comment]
I tried your changes with the [Backbone example](http://todomvc.com/examples/backbone/) and the template and now the text goes over the checkbox.

screen shot 2017-03-15 at 13 54 44

Can you please test your changes against the template or the Backbone example? I say this in the kindest way possible, but I could not care less about this change and there are so many other things I rather do than go back-and-forth on this PR :)

**Comment by patrickkettner** [comment]
apologies - that did not occur in `typescript-backbone`. updated accordingly

**Comment by sindresorhus** [comment]
Thank you! Glad to see this finally merged :)

**Comment by patrickkettner** [comment]
🎉 🎉 🎉
